netl.8
.\" @(#)man page for netl network logger/sniffer
.\" copyright 1997 graham the ollis
.ad b
.TH netl 8 "3 june 1997" "CORE software" "CORE software"
.AT 3
.de sh
.br
.ne 5
.PP
\fB\\$1\fR
.PP
..
.SH NAME
netl \- network logger/sniffer
.SH SYNOPSIS
.B netl
[
.B -v-
] [
.B -r-
] [
.BR \-f filename
] [
.B -z
] [
.BR \-o filename
] [
.B -h
] ['config line/rule'] [...]
.SH DESCRIPTION
.I netl
is a network logger and sniffing tool to aid a network administrator in
monitoring and analyzing incoming network packets.
.I netl
reads a config file (usually /etc/netl.conf see below) which defines the
type of packets that should be logged/sniffed, forks and runs as a
background process.  output is sent to syslogd, usually this  in turn sends the
data right to /var/adm/messages.  the other option is to have the packet
dumped to a file so that you can analize it with 
.BR neta (1)
which knows what fields are in an IP packet and will display
them for you.
.PP
.I netl
also listens to
.BR dcp (1)
packets if so specified in the config file.  so far the only 
.BR dcp (1)
packet
.I netl
will respond to is 
.I netl:readconfig
which will cause netl to dump the current configeration (and IP resolution
cache) so that the config file can be re-read.  this is handy if you make
modifictions to the config file while
.I netl
is running.
.PP
if you include config lines on the command line, then no config files
will be read.  this is a handy way to test netl config lines.  DO NOT use
regular hostnames on the command line.  use the dotted decimal, as no 
hostname resolution is done here (or in the config file either, use
.I alias
instead).
.SH OPTIONS
all boolean options can be postpended with a \+ or \- in order turn them
on or off.  the default invokation is always +, but the default value may
vary.
.PP
.TP 8
.BR \-f filename
use an alternative config file.
.PP
.TP
.BR \-o filename
send output to filename in addition to syslogd/stdout.  this option is usually 
available on non unix systems only.
.PP
.TP
.BR \-h
help.  print out usage information.
.PP
.TP
.BR \-r [+|-]
resolve IP numbers to their hostname (default is on).
.PP
.TP
.BR \-v [+|-]
display version number and copyright information (default is on).
.PP
.TP
.BR \-z [+|-]
run netl in the foreground, and send output to standard output
(default is off).
.SH CONFIG FILE
the 
.I netl 
distribution comes with a sample netl.conf file.  you should
use this as a template and then use the -z -f options in order to test it.
you can add comments to the config file by prepending the hash (#) charater
to the begining of the comment.  in general there are two parts of the config
file, although the borders aren't patrolled or anything, you can pretty much
put config lines anywhere in the file you like.  note that only one pass
is done to the config file, so 
.I device 
and 
.I alias
lines should probably go first.
.PP
.TP
\fBdevice\fR device type
this line specifies the type of device.  at the moment, only ethernet is
supported, so 
.I type
should be ethernet and 
.I device 
should be eth0, eth1, etc.
.PP
.TP
\fBdetect\fR
this directive tells netl to detect the hostname of the machine it is running
on and aliases it to the IP address.  also, local is aliased to this, though
lookups on the IP address will return the hostname so your log files will
be reasonable.
.PP
.TP
\fBalias\fR name ipnumber
this defines a alias for an ip numbers resolution.  this alias is used
for the config file, and for logging output.
.PP
action protocol [requirement]
.br
action is one of:
.TP
.I log
log the datagram in syslog
.TP
.I dump
save the entire datagram to a file in /tmp/netl
(make sure the directory /tmp/netl exists)
.TP
.I ignore
ignore all other requirement lines with the given values 
on this line.
.PP
protocol is one of:
.TP
.I tcp, udp, icmp
.PP
here are the rule fields. with the exception of name= you can put an !
before the field to negate it.  (e.g. \fB!flags=\fRfin) 
.PP
.I note:
you can only use each rule once on each line.  if you want to exclude more
than one IP (or other attribute multiple times)  you must use \fBignore\fR. if
you want to include more than on IP (or other attribute multiple times)
you will need to duplicate the line with a different IP number, and
everything else being the same. 
negation counts as a seperate
rule class so you can have (for example) both 
\fBsrcip=\fR and \fB!srcip=\fR 
on the same line.
.PP
.I another note: 
\fBflags=\fR is an exception.  you can use that multiple times
for each line.  that's sort of a hack though, don't use it unless you
have to as it may not be supported in later versions
.PP
.TP
\fBname=\fR
name isn't a real requirement.  it is simply a string which will be 
used to identify the datagram.  with log, this will show up as the 
first part of the log (i.e. when name=syn your output will be like
"netl: syn fred.com:77 => derf.org:88") or as part of the file name
of the dump.
.PP
.TP
\fBflags=\fR [TCP ONLY]
allows you to require certain tcp flags to be set.  you can specify
multiple flags by seperating them with commas (,).  here are the
flags you can specify: syn fin rst psh ack urg
.PP
.TP
\fBdstport=\fR [TCP and UDP ONLY]
specify the destination port to be used.  this can be either a
numerical port specification or a named service like telnet or ftp
as defined in /etc/services
.PP
.TP
\fBsrcport=\fR [TCP and UDP ONLY]
same as dstport except this is for the source port.
.PP
.TP
\fBdstip=\fR
specify the destination IP address.  this must either be a ip dotted 
decimal, or a host you have defined as an alias in this file.
note that netl WILL NOT do a hostname resolution for hosts in this config
file so make sure that any hostnames you use are defined before using
this rule.
.PP
.TP
\fBsrcip=\fR
same as dstip except this is for the source address.
.PP
.TP
\fBtype=\fR [ICMP ONLY]
type for icmp packets.  most useful thing here is "echo" (numerical
types can also be used.  there are all sorts of different icmp types
so if you are really interested, take a look at the source code:
lookup.c contains the lookup table.  the linux kernel also has some
data structures which may be helpful (usually you can find this in 
/usr/include/linux/icmp.h ).
.PP
.TP
\fBcode=\fR [ICMP ONLY]
this is sort of a sub type for icmp.  see the source code to this or the
kernel for all the answers you desire.
.SH SEE ALSO
.BR dcp (1),
.BR xd (1),
.BR neta (1),
.BR hwpassive (8)
.SH BUGS
almost certainly.  please send bug reports/fixes to me 
Graham THE Ollis <ollisg@ns.arizona.edu>
.SH COPYING
Copyright 1996, 1997 Graham THE Ollis
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.SH AUTHOR
Graham THE Ollis <ollisg@ns.arizona.edu>


HISTORY
version .95.2
-------------

 . mostly internal stuff.  added parse.c and grab.c modules, which contain
   code that used to be in the netl.c main module.  this is so that all the 
   network specific code is in grab.c, the true nightmare for porting is
   in this file... so if i can port grab.c to win32... then i'm done.

 . the \ character now works in the config file, if you want to have
   multiple line rules.  very unix-esq.

version .95
-----------

 . created man page documentation.

 . added `detect' config line, see the netl(8) manpage for details.
   this will reduce the amount of configeration you will have to do
   in the netl.conf file.

 . modified dist version of netl.conf so that it's more portable.
   in particular, i put in literal port numbers rather than using
   port names.

 . dcp.pl uses perl native Socket module instead of using net cat.
   `dcp.pl -o' uses the old method.

version .94
-----------

 . changed all the struct declerations to typedef declerations.  get off my
   back ok?  i've been programming perl for so long, forgot some of that
   strange C language. 

 . no longer depending on linux .h files for IP (and sub) headers.  this
   is for portability reasons.  if you are doing a port, make sure you get
   the endian type right!  there are now three global headers:
	global.h - general global stuff (endian integer conversion etc.)
	ether.h - ethernet specific data types, #defines
	ip.h - ip protocol specific data types, #defines
   logically, if your adding an interface (ATM, token ring, ppp, etc.)
   you'd want to start in ether.h.  if your adding an OS, look at global.h
   first.  if you want to add a protocol (IPX, etc.) start with ip.h and
   modify/add to these files.  if your trying to do all of this at once,
   god help you. 

 . fixed the clearipcache() function.  thus now, when you send a 
   netl:readconfig, netl will clear the ip cache before reading the file,
   avoiding all sorts of potential duplicates.  though you probably have
   an enourmous amount of memory like me and don't care... 

 . added some global options with ports to win32 in mind.  NO_SYSLOGD, if 
   defined in global.h will not even compile with the syslogd library stuff.  
   this is a good idea if your running from a non unix system.  note that 
   syslogd is really nice in that it partially compresses lines by indicating 
   repeated lines.  this won't save you if you get strobed, but will keep your 
   log file neat if some one ping -f's you from only one site.

 . added NO_TEEOUT option which is defined for unix.  the -o option should 
   only be necessary in win32 only.  even then, you should only compile with
   this option if your thinking of using it.  -o copies output to the specified
   file.

0.93 (and 0.92)
---------------

 a. -z option allows you to run netl in the foreground.  output is sent to
    stdout instead of syslogd.

 b. you can include configeration lines in the command line.  when you do
    this, the config file /etc/netl.conf will NOT be read.  note that (a)
    and (b) are meant to be used in concert with each other for testing
    and quick monitoring. 

 c. dcp is partially implemented.  netl will listen to port 47 udp
    (silently) if you `listen 47' in your config file.  when you send the
    message `netl:readconfig' netl is suppose to re read your config file. 
    dcp.pl is a perl script (you'll need net cat for this to work) which
    sends this message.  what is left to do for dcp?
	1. generic server side dcp interface libdcp.a file or somesuch.
	2. more solid C version of dcp client.

 d. netl is now officially under the GPL, see COPYING for more information.

 e. icmp codes are now included in some .h file somewhere rather than
    depending on your linux being overly up to date.

 f. neta no longer swaps source/destination ip/ethernet addresses
    inapropriately. 

 g. started work on hwpassive, which is no where near complete.

 h. tdr.pl is no included with netl distribution for convenience.  get
    your own version at: 

http://www.u.arizona.edu/~ollisg/core.html
	[ CORE ]

netl 0.91
---------

 minor updates

netl 0.90
---------

 initial release


netl.c
/*==============================================================================
| netl
|   optimized (and debugged) by Graham THE Ollis <ollisg@wwa.com>
|
|   Copyright (C) 1997 Graham THE Ollis <ollisg@wwa.com>
|
|   This program is free software; you can redistribute it and/or modify
|   it under the terms of the GNU General Public License as published by
|   the Free Software Foundation; either version 2 of the License, or
|   (at your option) any later version.
|
|   This program is distributed in the hope that it will be useful,
|   but WITHOUT ANY WARRANTY; without even the implied warranty of
|   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
|   GNU General Public License for more details.
|
|   You should have received a copy of the GNU General Public License
|   along with this program; if not, write to the Free Software
|   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
|
|  Date       Name	Revision
|  ---------  --------  --------
|  01 Feb 97  G. Ollis	modified, commented (and debugged)
|  08 Feb 97  G. Ollis	added IP address resolving.
|  23 Feb 97  G. Ollis	combined all network monitoring in to single program
|  28 Feb 97  G. Ollis	.92 added the -z option [ and the log() function to
|			replace syslog()]
|  05 Mar 97  G. Ollis	.93 added run time comunication.
|			took all direct syslog stuff out of this module.
|  07 Mar 97  G. Ollis	changed dump name to /tmp/netl/name-pid-time.dg.
|  26 sep 97  G. Ollis	took the networking specific portion of the code out
|			of this module and put it in to grab.c.
|=============================================================================*/

char	*id = "@(#)netl by graham the ollis <ollisg@wwa.com>";

#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <netdb.h>
#include <time.h>

#include "netl/global.h"
#include "netl/ether.h"
#include "netl/ip.h"

#include "netl/netl.h"
#include "netl/sighandle.h"
#include "netl/io.h"
#include "netl/options.h"
#include "netl/filter.h"
#include "netl/action.h"
#include "netl/config.h"
#include "netl/resolve.h"
#include "netl/grab.h"
#include "netl/check.h"
#include "netl/compiler.h"

/*==============================================================================
| GLOBALS
|=============================================================================*/

int (*grab)(char *buf);

char *out_path = "out";
char *filt_path = "filt";

// grab globals
char *in_path = "in";
void *grab_module = NULL;
int (*grab)(char *);

/*==============================================================================
| it's the clean up function!  it really doesn't need to do much so...
| (btw- clo is the name of the planet the decepticons invaded shortly after 
| the battle with unicron.  the autobots initially sustained incredable 
| losses, optimus prime returns and turns the tide with the help of the 
| "last autobot".  however, this has nothing to do with the clean up function)
|=============================================================================*/

void cleanup()
{
	clo();
}

/*==============================================================================
| int main(int, char **)
|=============================================================================*/

int
main(int argc, char *argv[])
{
#ifndef NO_SYSLOGD
	pid_t		temp;
#endif

	prog = argv[0];

	//setservent(TRUE);
	parsecmdline(argc, argv); 
	if(displayVersion) {
		fputs("netl ", stdout);
		puts(COPYVER);
	}

	preconfig();
	if(argc != 1) 
		while(--argc > 0) {
			argv++;
			if(argv[0][0] != '-') {
				parseconfigline(argv[0]);
				configfile = NULL;
			}
		}

	if(configfile != NULL)
#ifdef NO_SYSLOGD
		readconfig(configfile);
#endif
#ifndef NO_SYSLOGD
		readconfig(configfile, TRUE);
#endif
	postconfig();

	if(debug_mode) {
		//printconfig();
		return 1;
	}

	if(output_mode == OUT_MODE_C) {
		FILE *fp;
		if(output_name[0] == '-' && output_name[1] == 0)
			fp = stdout;
		else
			fp = fopen(output_name, "w");
		if(fp == NULL) {
			fprintf(stderr, "%s: error opening %s for write!\n",
				prog, output_name);
		}
		generate_c(fp);
		fclose(fp);
		return 0;
	}

#ifndef NO_SYSLOGD
	if(noBackground)
#endif
		return netl(netdevice);
#ifndef NO_SYSLOGD
	else {
		if((temp = fork()) == 0) 
			return netl(netdevice);

		if(temp == -1) {
			fprintf(stderr, "%s: unable to fork\n", prog);
			return 1;
		}
	}
#endif

	return 0;
}

/*==============================================================================
| prepare dl wrapper
|=============================================================================*/

void
prepare(char *dev)
{
	char str_buffer[1024];
	void (*prepare_function)(char *);

	if(grab_module != NULL) {
		nmclose(grab_module);
	}
	snprintf(str_buffer, 1024, "%s/%s/%s.so", so_path, in_path, grab_module_name);
	grab_module = nmopen(str_buffer);

	prepare_function = nmsym(grab_module, "prepare");

	prepare_function(dev);

	grab = nmsym(grab_module, "grab");
}

/*==============================================================================
| void netl(char *)
|=============================================================================*/

int reload_config_file = 0;

int
netl(char *dev)
{
	int		l;
	unsigned char buf[4096];

	ope("netl");
	log("starting netl, logging %s", dev);
	handle();

	prepare(dev);

	/*============================================================================
	| Entering the data collection loop
	|===========================================================================*/

	while(47) {			/* valnumdez's NoOp */
		if((l = grab(buf)) < 0) {
			log(strerror(errno));
			err("Error receiving RAW packet");
		} else {
			check(buf, l);
		}

		if(reload_config_file) {
			reload_config_file = 0;

			clearipcache();
			log("old ip cache cleared");

			clearconfig();

			preconfig();
			#ifdef NO_SYSLOGD
				readconfig(configfile);
			#else
				readconfig(configfile, noBackground);
			#endif
			postconfig();
			log("re-read configfile %s", configfile);
		}
	}

	return 0;
}
